### PR TITLE
Improve FieldFetcher retrieval of fields

### PR DESCRIPTION
**Pull request by cbuescher**
Currently FieldFetcher stores all the FieldContexts that are later used to
retrieve the fields in a List. This has the disadvantage that the same field
path can be retrieved several times (e.g. if multiple patterns match the same
path or if similar paths are defined several times e.g. with different formats).
Currently the last value to be retrieved "wins" and gets returned. We might as
well de-duplicate the FieldContexts by using a map internally, keyed by the
field path that is going to be retrieved, to avoid more work later.

### CODE REVIEW

**Review by jtibshirani — COMMENTED** [review]
Thanks, this is a nice simplification.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java [hunk 1]
```diff
@@ -48,7 +48,7 @@ public static FieldFetcher create(QueryShardContext context,
                                       SearchLookup searchLookup,
                                       Collection<FieldAndFormat> fieldAndFormats) {
 
-        List<FieldContext> fieldContexts = new ArrayList<>();
+        Map<String, FieldContext> fieldContexts = new HashMap<>();
         List<String> unmappedFetchPattern = new ArrayList<>();
         Set<String> mappedToExclude = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:53]: Sure, great unintended side effect

**Comment by cbuescher** [comment]
@jtibshirani thanks for the review, I pushed the changes you requested.

**Comment by cbuescher** [comment]
@elasticmachine update branch

**Comment by cbuescher** [comment]
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/eql-correctness

**Review by jtibshirani — APPROVED** [review]

**Pull request merged**